### ISSUE TITLE
BatchGetItem action in Step Function does not work

### ISSUE THREAD

**Issue by inaun** [aws/aws-sdk-js-v3#5432]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

When attempting to use BatchGetItem action in Step Function, Workflow Studio produces an error when "Key" from the boilerplate JSON is replaced with an actual table name. Error produced: "Errors detected in API Parameters.1) Parameters.RequestItems.Key : Value is mandatory. 2) Parameters.RequestItems.Key.Keys : Value is mandatory."

Attempting to execute the workflow produces a failure with error "The provided key element does not match the schema (Service: DynamoDb, Status Code: 400".  A valid schema is used as documented in BatchGetItem API documentation at https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchGetItem.html.

### SDK version number

Not applicable -- state machine

### Which JavaScript Runtime is this issue in?

Browser

### Details of the browser/Node.js/ReactNative version

AWS Workflow Studio

### Reproduction Steps

1) Create a new State Machine, 
2) add BatchItemGet action, 
3) configure API parameters with table name and keys

### Observed Behavior

Errors produced with valid JSON as documented for BatchGetItem API

### Expected Behavior

Successful execution

### Possible Solution

Fix State Machine (Step Function) action BatchGetItem to correctly interact with BatchGetItem API

**Comment by RanVaknin** [comment]
Hi @inaun ,

Thanks for opening this issue. From what I can tell there are two issues at hand here:

>"Errors detected in API Parameters.1) Parameters.RequestItems.Key : Value is mandatory. 2) Parameters.RequestItems.Key.Keys : Value is mandatory."

This implies that, even if you did specify the table name (which would replace "Key"), you're missing the "Keys" part, which defines the primary keys of the items you want to retrieve from that table.

>"The provided key element does not match the schema (Service: DynamoDb, Status Code: 400".

This error can is a service error, it is usually thrown when the request you are sending contains fields that are either incorrect in type / name or simply do not conform to the table structure you have specified when you created the table. 

Without seeing your table schema, or your code, this will be impossible to reproduce. 

Additionally, you mentioned this is related to a state machine / step functions, however I'm not sure how this problem is specific to that platform. Did you try running your code from a local node.js application? 

You can share the schema of your table by copying the contents of the following CLI command:

```console
$ aws dynamodb describe-table --table-name YourTableName
```

And also share the actual payload you are sending so I can better assist you. Just make sure to redact any sensitive information.

Thanks,
Ran~

**Comment by inaun** [comment]
I'm sorry, but it does not imply that I left out "keys". As you point out, the first error referrs to "key", and is the primary problem. The second error is a result of the first error, not the root issue. I further stated that I provided a correct schema as per the AWS documentation for the BatchGetItems API, which I linked to. If I may be so bold, it's quite rude to automatically assume I left out vital parts of the JSON even after I provided a link to the documentation. Did you attempt to reproduce the error before concluding I don't know enough to provide the actual search keys I'm querying on?

As to the JSON I configured for API Parameters for the Step Function action BatchGetItems, here it is. We've gotten off on the wrong foot due to the rather arrogant way you responded. Let's see if maybe we can find our way back to problem solving. Maybe I do have a bug in my JSON, but leaving out an entire section is definitely not the problem.

{
   "RequestItems": { 
      "reportsPOCddbt" : { 
         "Keys": [ 
            { 
               "PK": {"S":"A#RID2047#"},
               "SK": {"S":"OID321886"}
            },
            { 
               "PK": {"S":"A#RID2312#"},
               "SK": {"S":"OID321886"}
            }
         ],
         "ProjectionExpression": "Name, Date"
      }
   }
}

**Comment by inaun** [comment]
PS, I mention State Machine (aka Step Function) because this problem is directly related to that service. It appears that the reported State Machine Action is not interacting correctly with the DynamoDB API. If I've reported this bug in the wrong area, please feel free to point me to the correct spot. Thx.

**Comment by RanVaknin** [comment]
Hi @inaun,

Oh wow, Im sorry if I came across as arrogant, that was not at all the intention. I think this was a mix of misunderstand on my part on the issue at hand, in conjunction with not getting all the information in the intake form. The reality is that we get questions from developers in every level of their career, and in many of the cases DynamoDB issues that are opened are due to misconfiguration of the request.

This was not to assume that you don't know what your own code does, but with the lack of schema and/or actual request code there was not a lot for me to go off. And the first step was for me to rule out the obvious by taking the error at face value.

I actually spent a fair bit of time trying to reproduce this before realizing that you are using the **_console_** and not the SDK (this github repo is aimed helping SDK customers). Typically support for the console customers needs to go through the AWS console and file a support ticket, but since you are here It's important for me to unblock you. I spent a little more time on the console and got it to work. 

When using the editor I saw the same error:
>"Errors detected in API Parameters.1) Parameters.RequestItems.Key : Value is mandatory. 2) Parameters.RequestItems.Key.Keys : Value is mandatory."

I think you are right in your initial assumption that this is a bug. Probably some console side validations that are incorrectly defined. I was able to work around it by going into the code tab and running my code from there:

![image](https://github.com/aws/aws-sdk-js-v3/assets/50976344/0bff2e40-923b-4d0e-a457-cf3d08e84c0c)

This resulted in successful invocation and a valid response:

![image](https://github.com/aws/aws-sdk-js-v3/assets/50976344/4fb8a513-0e9a-4ca2-96f7-4259856f423e)


I want to emphasize  that I did try to reproduce this both using the SDK from a node environment (which went through without issues) in addition to trying to setup a step function but the Step function API rejects the creation of a state machine stating that `BatchGetItem` arn is incorrect:

```cli
 $ aws stepfunctions create-state-machine \
  --name "YourStateMachineName" \          
  --role-arn "arn:aws:iam::REDACTED:role/StepFunctions-DynamoDB-Role" \
  --definition '{     
      "Comment": "A simple AWS Step Functions state machine that retrieves items from DynamoDB.",
      "StartAt": "RetrieveItems",
      "States": {
        "RetrieveItems": {
          "Type": "Task",                 
          "Resource": "arn:aws:states:::dynamodb:batchGetItem",         
          "Parameters": {
            "RequestItems": {
              "5432": {
                "Keys": [
                  {
                    "PK": {"S": "A#RID2047#"},
                    "SK": {"S": "OID321886"}
                  },
                  {
                    "PK": {"S": "A#RID2312#"},
                    "SK": {"S": "OID321886"}
                  }
                ],
                "ProjectionExpression": "#N, #D",
                "ExpressionAttributeNames": {"#N": "Name", "#D": "Date"}
              }
            }
          },
          "End": true
        }
      }
    }'


An error occurred (InvalidDefinition) when calling the CreateStateMachine operation: Invalid State Machine Definition: 'SCHEMA_VALIDATION_FAILED: The resource provided arn:aws:states:::dynamodb:batchGetItem is not recognized. The value is not a valid resource ARN, or the resource is not available in this region. at /States/RetrieveItems/Resource'
```

[Their documentation ](https://docs.aws.amazon.com/step-functions/latest/dg/connect-ddb.html)also does not list BatchGetItem as one of the supported transformations between step function and dynamo. So I have no idea how it works from the console if their API server and API docs don't support it. 

In conclusion, I hope this unblocks you and I apologize again for the misunderstanding. If you have any other questions please let me know and I'll be happy to help. 

All the best,
Ran~
